### PR TITLE
fast path next for String type

### PR DESCRIPTION
**Pull request by vtjnash**
ref parse_int test in #16128

### CODE REVIEW

File: base/strings/string.jl [hunk 1]
```diff
@@ -59,31 +59,42 @@ function next(s::String, i::Int)
...
         throw(UnicodeError(UTF_ERR_INVALID_INDEX, i, d[i]))
     end
-    trailing = utf8_trailing[b+1]
+    trailing = utf8_trailing[b + 1]
     if length(d) < i + trailing
         return '\ufffd', i+1
     end
     c::UInt32 = 0
-    for j = 1:trailing+1
+    for j = 1:(trailing + 1)
```
Comment (ararslan) [inline base/strings/string.jl]: Is this just a style thing or will the parentheses be required at some point?

Comment (vtjnash) [inline base/strings/string.jl]: style 😎

File: base/strings/string.jl [hunk 2]
```diff
@@ -59,31 +59,42 @@ function next(s::String, i::Int)
...
+    return slow_utf8_next(d, b, i)
 end
 
 function first_utf8_byte(ch::Char)
     c = UInt32(ch)
-    c < 0x80    ? c%UInt8 :
-    c < 0x800   ? ((c>>6)  | 0xc0)%UInt8 :
-    c < 0x10000 ? ((c>>12) | 0xe0)%UInt8 :
-                  ((c>>18) | 0xf0)%UInt8
+    b = c < 0x80    ? c%UInt8 :
```
Comment (ararslan) [inline base/strings/string.jl]: Sorry if this is a dumb question but I don't know much about this stuff. Does assigning to a variable first not cause any extra allocations or anything?

Comment (vtjnash) [inline base/strings/string.jl]: no, variables just give the values a name. I just added this one here to make it more visible to future readers that this loop is computing a return value

Comment (ararslan) [inline base/strings/string.jl]: Okay, thanks for the explanation, I appreciate it. Still learning.

**Comment by nalimilan** [comment]
Clever!

**Comment by tkelman** [comment]
I guess (part of?) the comment can be deleted now.

**Pull request merged**